pulsar-broker/src/main/java/org/apache/pulsar/broker/web/AuthenticationFilter.java

```diff
@@ -42,11 +43,18 @@
     private static final Logger LOG = LoggerFactory.getLogger(AuthenticationFilter.class);
 
     private final AuthenticationService authenticationService;
+    private final boolean isBasic;
 
     public static final String AuthenticatedRoleAttributeName = AuthenticationFilter.class.getName() + "-role";
 
     public AuthenticationFilter(PulsarService pulsar) {
         this.authenticationService = pulsar.getBrokerService().getAuthenticationService();
+        if (pulsar.getConfiguration().getAuthenticationProviders()
+                .contains(AuthenticationProviderBasic.class.getName())) {
+            isBasic = true;
```

Comment:
Rather than a ad-hoc solution, is it possible to have a general way to trigger the right header in the response?

Reply:
Maybe we can modify `AuthenticationService::authenticateHttpRequest`. Currently we just ignore exceptions in the method to try all auth providers, but probably we should gather information from the providers. Because "WWW-Authenticate" header can contain multiple challenges, we can suggest all supported authentication types at the first 401 response.
https://tools.ietf.org/html/rfc7235#section-4.1

Also, there is (at least) another response header,  "Authentication-Info" (https://tools.ietf.org/html/rfc7615), that will carry information when authentication succeeds. So it would be better to design the API for this mechanism so that providers can respond with arbitrary headers and/or body data regardless of authentication results.

https://tools.ietf.org/html/rfc7235#section-2.1
> HTTP does not restrict applications to this simple challenge-response
   framework for access authentication.  Additional mechanisms can be
   used, such as authentication at the transport level or via message
   encapsulation, and with additional header fields specifying
   authentication information.  However, such additional mechanisms are
   not defined by this specification.

Reply:
I think modifying `authenticateHttpRequest` is not efficient because the information is gathered every time authentication failed.
I fixed AuthentcationFilter class to gather information from providers when it is initialized.

Reply:
I don’t think we are always be able to gather information on initialization. For example, Digest authentication needs to generate a challenge (nonce) every time.

Reply:
Make sense.
I added PulsarHttpAuthenticationException class, which has realm information for `WWW-Authentication` header and the information is gathered in `authenticateHttpRequest`.